### PR TITLE
apiextensions: forbid false x-kubernetes-preserve-unknown-fields

### PR DESCRIPTION
**Pull request by sttts**
This is needed so we can catch/forbid explicit false values from the user to avoid confusion

Compare https://github.com/kubernetes/enhancements/pull/1061.

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go [hunk 1]
```diff
@@ -66,6 +66,8 @@ type Extensions struct {
 	// in the validation schema. This affects fields recursively,
 	// but switches back to normal pruning behaviour if nested
 	// properties or additionalProperties are specified in the schema.
+	// False means that the pruning behaviour is inherited from the parent.
+	// False does not mean to activate pruning.
 	XPreserveUnknownFields bool
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go]: am undecided whether this is a good idea. In the structural schema we never want to see the nil case. Hence, I define the false differently.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/convert.go [hunk 2]
```diff
@@ -241,8 +241,12 @@ func newExtensions(s *apiextensions.JSONSchemaProps) (*Extensions, error) {
 		return nil, nil
 	}
 
+	if s.XPreserveUnknownFields != nil && *s.XPreserveUnknownFields == false {
+		return nil, fmt.Errorf("'x-kubernetes-preserve-unknown-fields' must be true or undefined")
+	}
+
 	return &Extensions{
-		XPreserveUnknownFields: s.XPreserveUnknownFields,
+		XPreserveUnknownFields: *s.XPreserveUnknownFields,
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/convert.go]: can be nil here, this is an npe, right?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 3]
```diff
@@ -194,8 +194,8 @@ func ConvertJSONSchemaPropsWithPostProcess(in *apiextensions.JSONSchemaProps, ou
 		}
 	}
 
-	if in.XPreserveUnknownFields {
-		out.VendorExtensible.AddExtension("x-kubernetes-preserve-unknown-fields", true)
+	if in.XPreserveUnknownFields != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: `&& *in.XPreserveUnknownFields`?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/convert.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/convert.go]: fixed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go]: Validation enforces `true` if non-nil. Doesn't make much of a difference. In my variant even the non-validating `false` goes through.

**Comment by liggitt** [comment]
roundtrip and openapi tests look out of date

File: cmd/genkubedocs/gen_kube_docs.go [hunk 4]
```diff
@@ -54,7 +53,7 @@ func main() {
 	switch module {
 	case "kube-apiserver":
 		// generate docs for kube-apiserver
-		apiserver := apiservapp.NewAPIServerCommand(server.SetupSignalHandler())
+		apiserver := apiservapp.NewAPIServerCommand()
```
Comment (liggitt) [inline cmd/genkubedocs/gen_kube_docs.go]: are these changes supposed to be in this PR?

**Pull request merged**